src/transport/PointToPointBroker.cpp

```diff
@@ -566,10 +672,74 @@ std::vector<uint8_t> PointToPointBroker::recvMessage(int groupId,
 
     Message message = recvEndpoints[label]->recv();
 
-    // TODO - possible to avoid this copy?
     return message.dataCopy();
 }
 
+std::vector<uint8_t> PointToPointBroker::recvMessage(int groupId,
+                                                     int sendIdx,
+                                                     int recvIdx)
+{
+    // If we don't need to receive messages in order, return here
+    if (!orderMsg.load(std::memory_order_acquire)) {
+        return doRecvMessage(groupId, sendIdx, recvIdx);
+    }
+
+    int expectedSeqNum = getExpectedSeqNum(groupId, sendIdx);
+    std::vector<uint8_t> recvData;
+    std::vector<uint8_t> returnData;
+
+    // Loop to receive the right sequence number
+    while (true) {
+        // First, receive from the transport layer
+        recvData = doRecvMessage(groupId, sendIdx, recvIdx);
+
+        // Second, get the sequence number from the received message
+        int seqNum;
+        faabric::util::readBytesOf(
+          recvData, recvData.size() - sizeof(int), &seqNum);
+        recvData.resize(recvData.size() - sizeof(int));
+
+        // If the sequence numbers match, exit the loop
+        if (seqNum == expectedSeqNum) {
+            returnData = std::move(recvData);
+            break;
+        }
+
+        // If not, we must insert the received message in the out of order
+        // received messages. We must also check if the sequence number we want
+        // is in the list.
+
+        // First insert
```

Comment:
Fast path:

- Avoid message copy by taking ownership.
- Do `outOfOrderMsgs.emplace_back(message)`.
- Retrieval code will have a special case, `if(outOfOrderMsgs.back().seqNum == expectedSeqNum) { // it's the one we want }`.
- Otherwise do a simple iteration over the list, check each `msg.seqNum`.